adaptdl/adaptdl/torch/parallel.py

```diff
@@ -59,9 +67,13 @@ def __init__(self, model, optimizer, lr_scheduler=None, mp_scaler=None,
         for param in model.parameters():
             param.register_hook(functools.partial(self._backward_hook, param))
 
-        # Setup for AdaScale, must be after registering backward hooks!
-        self.adascale = AdaScale(self, optimizer, mp_scaler=mp_scaler,
-                                 patch_optimizer=True)
+        # Setup for the scaling_rule, must be after registering backward hooks
+        # because some of them need to register their own backward hooks.
+        self.gns = GradientNoiseScale(self, optimizer, mp_scaler=mp_scaler)
+        self.scaling_rule = scaling_rule or AdaScale()
+        self.scaling_rule.adp = self
```

Comment:
Can this be moved to `register_optimizer` (and maybe rename that as `register_scaling_rule`)?